### PR TITLE
Upgrade buildbox to go 1.17.7 & tag as `teleport10`

### PR DESCRIPTION
**Pull request by tcsc**
Prior to this patch the teleport buildbox version has been tagged with the Go version for the current release. This bit us during the Teleport 9 development cycle, as both Teleport 8 and 9 use the same version of Go but require different versions of Rust, and we were unable to distinguish between the 2 buildbox versions. 

At the time, Teleport 8 was individually patched to create a new `teleport8` buildbox tag, decoupling the buildbox version from the Go version. This was never ported into `master` and now we find the teleport 9 branch sharing the same buildbox tag as `master`.

This patch forward-ports all the changes made to `branch/v8` and updates them for `master`, creating a new `teleport10` buildbox tag. The idea is that we will create a new tag for `teleport11` at the same time the release branch for Teleport 10 is mad at some point in the future.

Once this is merged, Drone will create and push new buildbox images, which will become available for CI. A subsequent patch will update the CI scripts to use the new `teleport10` buildbox images.

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -4242,7 +4256,8 @@ type: kubernetes
 name: build-docker-images
 
 environment:
-  RUNTIME: go1.17.2
+  BUILDBOX_VERSION: "teleport9"
```
Comment (zmb3) [inline .drone.yml:4259]: You mean `teleport10`, right?

**Review by zmb3 — CHANGES_REQUESTED** [review]

**Comment by wadells** [comment]
Philosophically, I'd rather we don't use floating images references at all.  Better to do something like go.sum and pin to a specific SHA.  I get a bit further into this idea here:

https://github.com/gravitational/teleport/issues/6895

Why?  Non-deterministic build inputs (in this case a floating OCI image ref) *will* eventually result in nondeterministic build outputs.  This is the 2nd or 3rd time we've been hit by a image tag changing and breaking our build.  While your implementation of locking a tag to a branch helps, it will still break building older version on that branch.  E.g. we can expect the tip of main or v9 to build, but not the commit from a month or two ago which now references a buildbox tag from the "future".

My sketch of an implementation would be:

1) allow floating OCI refs to be specified in dronegen -- for user friendliness
2) have dronegen dereference these to a fixed image sha set at the moment dronegen is run, and bake that sha into a lockfile.  Use the SHA in the pipeline.
3) When we want to update images, run a dronegen command to re-fetch one or more images and update the lockfile with new shas (just like `go get` updates go.sum).  We probably want this to be a separate command than the default `dronegen` because we don't want folks changing the file for other reasons to accidentally pick up (and have to test) newer version of images unrelated to their changes.

That said, this is definitely a step in the right direction, and I don't' want to let good get in the way of better.

File: dronegen/buildbox.go [hunk 2]
```diff
@@ -58,19 +58,21 @@ func buildboxPipelineStep(buildboxName string, fips bool) step {
...
-		"UID":     {raw: "1000"},
-		"GID":     {raw: "1000"},
+		"BUILDBOX_VERSION": buildboxVersion,
+		"UID":              {raw: "1000"},
+		"GID":              {raw: "1000"},
 	}
-	p.Trigger = triggerPushMasterOnly
+
+	// only on master for now; add the release branch name when forking a new release series.
+	p.Trigger = pushTriggerForBranch("master")
```
Comment (wadells) [inline dronegen/buildbox.go:75]: Would master and `branch/*` save us the need for changes later?

**Review by wadells — APPROVED** [review]

**Comment by tcsc** [comment]
> Philosophically, I'd rather we don't use floating images references at all. Better to do something like go.sum and pin to a specific SHA.

Right there with you. This is a step on the path that solves an immediate problem (multiple release branches stomping on each other) rather than the destination. 

I've started writing down my thoughts about what we should do [here](https://github.com/gravitational/teleport/pull/10206). Comments and suggestions more than welcome.

File: .drone.yml (hunk 1 above)
Comment (tcsc) [reply .drone.yml:4259]: Bah. Apparently `dronegen` doesn't automatically update that one. *sigh*

File: dronegen/buildbox.go (hunk 2 above)
Comment (tcsc) [reply dronegen/buildbox.go:75]: Yeah, not a bad idea.

**Review by nklaassen — COMMENTED** [review]
> The idea is that we will create a new tag for teleport11 at the same time the release branch for Teleport 10 is mad at some point in the future.

Should we add this to the pre/post release checklist, or better yet put it into `make setver` or some other script that we run when cutting a release branch?

**Review by zmb3 — APPROVED** [review]

**Comment by tcsc** [comment]
> Should we add this to the pre/post release checklist, or better yet put it into `make setver` or some other script that we run when cutting a release branch?

See #10689

**Pull request merged**